api/src/services/spatial-service.ts

```diff
@@ -99,6 +111,38 @@ export class SpatialService extends DBService {
     });
   }
 
+  /**
+   * Collect transforms from db, run transformations on submission id, save result to spatial component table
+   *
+   * @param {number} submissionId
+   * @param {string} spatialTransformName
+   * @return {*}  {Promise<void>}
+   * @memberof SpatialService
+   */
+  async runSpatialTransforms(submissionId: number): Promise<void> {
+    const spatialTransformRecords = await this.getSpatialTransformRecords();
+    let transformed;
+
+    spatialTransformRecords.forEach(async (transformRecord) => {
+      transformed = await this.spatialRepository.runSpatialTransformOnSubmissionId(
+        submissionId,
+        transformRecord.transform
+      );
+
+      transformed.forEach(async (dataPoint) => {
+        const submissionSpatialComponentId = await this.spatialRepository.insertSubmissionSpatialComponent(
+          submissionId,
+          dataPoint.result_data
+        );
+
+        await this.insertSpatialTransformSubmissionRecord(
+          transformRecord.spatial_transform_id,
+          submissionSpatialComponentId.submission_spatial_component_id
+        );
+      });
+    });
+  }
```

Comment:
We'll need to tweak this slightly.  

A `forEach` loop doesn't honour async await between loops.  It will just run as fast as possible, kicking off N number of calls, which internally will honour the async await, but externally will run in parallel and may finish out of order, etc.

So the `transformed` variable may actually get written/read in an unexpected order, the way it is set up currently.
Easy fix is just to initialize `transformed` inside the loop, rather than outside, so its self contained.
`let transformed = await this.spatia...`

Also, if it matters that these transformations run in order, then we'll need to use a different kind of loop.

Reply:
Fixed